### PR TITLE
LPS-151251 Create URLBuilder to prep converting HttpComponentsUtil param methods into builder pattern

### PR DESCRIPTION
**Pull request by julschong**
@tinatian Hi Tina,

this is resubmitted https://github.com/liferay-core-infra/liferay-portal/pull/927

Main Task: [LPS-150839 Remove petra-apache-http-components](https://issues.liferay.com/browse/LPS-150839)
Sub Task: [LPS-151251 Create URLBuilder interface to provide builder interface to HttpComponentUtil](https://issues.liferay.com/browse/LPS-151251)

URLBuilder has no logic currently. Only using HttpComponentsUtil's logic.
Next part for this task is to move the logic to URLBuilder and apply to usages.

Additional note for this pull:

1. URLBuilder now save each operation and only modify _url inside build() method. This change is to follow conventional Builder pattern for easier understanding.
2. I did change the logic so that each operation is removed after processing so that repearting build() will not cause an issue.
3. Added a test case to make sure operational sequence is followed

Thank you!

### CODE REVIEW

**Comment by julschong** [comment]
@catarineaguiar
Hi Catarine,
Could you please help me scan this test result?
Thank you!

**Comment by catarineaguiar** [comment]
@julschong no regressions found.

File: portal-kernel/src/com/liferay/portal/kernel/url/URLBuilder.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+		return this;
+	}
+
+	public String build() {
+		if (_url == null) {
+			return null;
+		}
+
+		while (!_urlOperations.isEmpty()) {
+			URLOperation urlOperation = _urlOperations.remove(0);
```
Comment (tinatian) [inline portal-kernel/src/com/liferay/portal/kernel/url/URLBuilder.java:76]: @julschong 
The remove(0) here means you want to remove the first one. Removing the first one is usually a key usage of a queue, it is not a general usage for List.

More importantly, removing while iterating is dangerous,  it is not always safe to do, please see this for more details: https://www.techiedelight.com/remove-elements-list-within-loop/

Please make this a normal for loop, apply the operations, then clear the list.

**Comment by tinatian** [comment]
@julschong Please fix and resend.

Tina.

File: portal-kernel/src/com/liferay/portal/kernel/url/URLBuilder.java (hunk 1 above)
Comment (julschong) [reply portal-kernel/src/com/liferay/portal/kernel/url/URLBuilder.java:76]: got it

**Pull request closed without merging**